src/profile-logic/symbol-store.js

```diff
@@ -361,82 +362,121 @@ export class SymbolStore {
+              option3Requests,
+              option3ResponsePromise,
+              errorMap,
+              successCb
             );
+
+          if (option4Requests.length === 0) {
+            // Done!
+            return;
           }
-        } catch (error) {
-          // There was a problem for the entire chunk of requests, for example there
-          // might have been a problem parsing the response JSON.
-          // Map all requests to the same error, and then try to get symbols from the browser.
-          const requestsWithErrors = requests.map((request) => ({
-            request,
-            error: new SymbolsNotFoundError(
-              'There was a problem with the JSON returned by the symbolication API.',
-              request.lib,
-              error
-            ),
-          }));
-          await this._getSymbolsFromBrowser(
-            requestsWithErrors,
+
+          // Some libraries are still remaining, try option 4 for them.
+          // Option 4 is symbolProvider.requestSymbolTableFromBrowser.
+          await this._getSymbolTablesFromBrowser(
+            option4Requests,
+            errorMap,
             demangleCallback,
             successCb,
             errorCb
           );
         }
-      })
+      )
     );
   }
 
-  // Try to get symbols from the browser for any libraries which couldn't be
-  // symbolicated with the symbolication API.
-  async _getSymbolsFromBrowser(
-    requestsWithErrors: Array<{|
-      // A symbolication request for one library.
-      request: LibSymbolicationRequest,
-      // The error which was encountered when trying to obtain symbols for
-      // this library using the symbolication API.
-      // This gets reported in the final error if all avenues fail.
-      error: Error,
-    |}>,
+  // This method awaits `responsesPromise`. All successful responses are forwarded
+  // to successCb.
+  // Returns any requests which were not successful.
+  // Mutates errorMap by adding the error that was encountered for a failed request.
+  async _processSuccessfulResponsesAndReturnRemaining(
+    requests: LibSymbolicationRequest[],
+    responsesPromise: Promise<LibSymbolicationResponse[]>,
+    errorMap: Map<LibSymbolicationRequest, Error[]>,
+    successCb: (RequestedLib, Map<number, AddressResult>) => void
+  ): Promise<LibSymbolicationRequest[]> {
+    try {
+      const responses: LibSymbolicationResponse[] = await responsesPromise;
+
+      const failedRequests = [];
+      for (let i = 0; i < responses.length; i++) {
+        const response = responses[i];
+        if (response.type === 'SUCCESS') {
+          // We've found symbols for this library!
+          const { lib, results } = response;
+          successCb(lib, results);
+        } else {
+          // No symbols for this library, it'll need to try the next option.
+          const { request, error } = response;
+          failedRequests.push(request);
+          (errorMap.get(request) ?? []).push(error);
```

Comment:
Ah, maybe I should add a comment. `request` is guaranteed to be in the `errorMap` - we initialize `errorMap` with all requests. The `?? []` is just there because it's the shortest way I could think of to appease Flow. In Rust I would use `.unwrap()` here to say "I'm sure that this exists and I'm willing to risk a panic if my reasoning about this code is wrong". In JS I could throw an exception, but that would be three extra lines.
Or maybe I should use `ensureExists` from the flow helpers? But that one takes an error message...

Reply:
I've added two comments.

Reply:
>  Or maybe I should use ensureExists from the flow helpers? But that one takes an error message...

The error message isn't mandatory, the tool adds a default one if it's missing.

Reply:
So that would have been my preference :-)